packages/vscode-extension/src/commonlib/m365Login.ts

```diff
@@ -0,0 +1,233 @@
+import { getDefaultString, localize } from "../utils/localizeUtils";
+import { AppStudioScopes } from "@microsoft/teamsfx-core";
+
+const SERVER_PORT = 0;
+const cachePlugin = new CryptoCachePlugin(m365CacheName);
+
+const config = {
+  auth: {
+    clientId: "7ea7c24c-b1f6-4a20-9d11-9ae12e9e7ac0",
+    authority: "https://login.microsoftonline.com/common",
+  },
+  system: {
+    loggerOptions: {
+      // @ts-ignore
+      loggerCallback(loglevel, message, containsPii) {
+        if (loglevel <= LogLevel.Error) {
+          VsCodeLogInstance.error(message);
+        }
+      },
+      piiLoggingEnabled: false,
+      logLevel: LogLevel.Error,
+    },
+  },
+  cache: {
+    cachePlugin,
+  },
+};
+
+export class M365Login extends BasicLogin implements M365TokenProvider {
+  private static instance: M365Login;
+  private static codeFlowInstance: CodeFlowLogin;
+
+  private constructor() {
+    super();
+    M365Login.codeFlowInstance = new CodeFlowLogin([], config, SERVER_PORT, m365CacheName);
+  }
+
+  /**
+   * Gets instance
+   * @returns instance
+   */
+  public static getInstance(): M365Login {
+    if (!M365Login.instance) {
+      M365Login.instance = new M365Login();
+    }
+
+    return M365Login.instance;
+  }
+
+  /**
+   * Get team access token
+   */
+  async getAccessToken(tokenRequest: TokenRequest): Promise<Result<string, FxError>> {
+    await M365Login.codeFlowInstance.reloadCache();
+    let tokenRes: Result<string, FxError>;
+    if (!M365Login.codeFlowInstance.account) {
+      if (tokenRequest.showDialog) {
+        const userConfirmation: boolean = await this.doesUserConfirmLogin();
+        if (!userConfirmation) {
+          ExtTelemetry.sendTelemetryEvent(TelemetryEvent.Login, {
+            [TelemetryProperty.AccountType]: AccountType.M365,
+            [TelemetryProperty.Success]: TelemetrySuccess.No,
+            [TelemetryProperty.UserId]: "",
+            [TelemetryProperty.Internal]: "",
+            [TelemetryProperty.ErrorType]: TelemetryErrorType.UserError,
+            [TelemetryProperty.ErrorCode]: `${getDefaultString(
+              "teamstoolkit.codeFlowLogin.loginComponent"
+            )}.${ExtensionErrors.UserCancel}`,
+            [TelemetryProperty.ErrorMessage]: `${getDefaultString(
+              "teamstoolkit.common.userCancel"
+            )}`,
+          });
+          return err(
+            UserCancelError(getDefaultString("teamstoolkit.codeFlowLogin.loginComponent"))
+          );
+        }
+        M365Login.codeFlowInstance.status = loggingIn;
+        this.notifyStatus(tokenRequest);
+      }
+      tokenRes = await M365Login.codeFlowInstance.getTokenByScopes(tokenRequest.scopes);
+      await this.notifyStatus(tokenRequest);
+    } else {
+      tokenRes = await M365Login.codeFlowInstance.getTokenByScopes(tokenRequest.scopes);
+    }
+
+    if (tokenRes.isOk()) {
+      return ok(tokenRes.value);
+    } else {
+      return tokenRes;
+    }
+  }
+
+  async getJsonObject(
```

Comment:
Does this mean we should write following logics where we need a sharepoint token?
1. get graphReadUser token
2. use the token to get spfx tenenat and obtain sharepoint token scope
3. use the scope to request sharepoint token

Reply:
The process is right. We can follow this code to create a spfx token.
```
const graphTokenRes = await M365TokenInstance.getAccessToken({scopes: GraphReadUserScopes});
  if (graphTokenRes.isOk()) {
    const spfxToken = M365TokenInstance.getAccessToken({scopes: SPFxScopes(await getSPFxTenant(graphTokenRes))});
  }
```